### PR TITLE
feat(mlx): quantization-aware training for LoRA on Apple Silicon

### PR DESCRIPTION
**Pull request by BardiaKoopah**
Implements `qat_scheme` for the MLX backend, replacing the `NotImplementedError`
  stub in `get_peft_model` (`unsloth_zoo/mlx/loader.py:7534`).

  ## The problem

  MLX's `merged_4bit` save does **not** leave the base quantization untouched.
  `mlx_lm.tuner.lora.LoRALinear.fuse(dequantize=False)` dequantizes the base, adds the
  LoRA delta, and then **requantizes**:

  ```python
  weight = mx.dequantize(weight, linear.scales, linear.biases, ...)   # 1. dequantize
  fused_linear.weight = weight + delta                                 # 2. add BA
  if is_quantized and not dequantize:
      fused_linear = nn.QuantizedLinear.from_linear(...)               # 3. REQUANTIZE
  ```

  So the weight that actually ships is `quantize(dequantize(W_q) + B @ A)`. But during
  normal QLoRA training the adapter is applied in **full precision** on top of the
  quantized base, so that final requantization is error the model never sees while
  training.

  It is not a rounding detail. Requantization is not idempotent even with a zero LoRA
  delta — `max |W_dq - fq| = 0.0037` on a 4-bit/group-64 base.

  ## Evidence

  Qwen2.5-0.5B-Instruct-4bit, wikitext-2, standard LoRA (r=16, all 24 layers),
  measured pre-fuse vs post-fuse:

  | steps | baseline degradation | QAT degradation |
  |---|---|---|
  | 0 | +0.0112 | −0.0002 |
  | 50 | +0.1133 | +0.0001 |
  | 150 | +0.1657 | +0.0000 |
  | 300 | +0.2063 | −0.0000 |

  The headline number is retention, not degradation. Over 300 steps:

  - **baseline** gains 0.306 nats while training, keeps **0.111** after the save →
  **36%**
  - **QAT** gains 0.241, keeps **0.241** → **~100%**

  64% of what LoRA learns never reaches the saved model. QAT ships a model **0.130 
  nats better** (perplexity 23.19 vs 26.42). The advantage widens with training: 0.058
  → 0.092 → 0.130 at 50/150/300 steps.

  Through the real `save_pretrained_merged(save_method="merged_4bit")` → reload path
  (Qwen2.5-0.5B, short overfit run, so the magnitude is exaggerated but the mechanism
  is the product path):

  | | before save | after reload |
  |---|---|---|
  | no QAT | 0.000087 | **0.930841** |
  | QAT | 0.004248 | **0.004422** |

  **Memory is free.** Peak with gradient checkpointing: baseline 1412 MB, QAT 1404 MB
  (24 LoRA layers, seq 512, 16GB M2). Checkpointing fully absorbs the ~1.6x transient
  overhead. Forward cost is ~2x.

  ## Design — and why this diverges from CUDA

  CUDA/torchao fake-quantizes the **frozen base** and leaves the adapter in high
  precision. I verified this by running `QATConfig(step="prepare")` on a LoRA'd
  `nn.Linear`: `base_layer` becomes `FakeQuantizedLinear`, the base weight gets no
  gradient, and the adapter trains through the quantizer.

  That is correct for bitsandbytes. It is the **wrong target for MLX**: simulating
  `quantize(W)` while `fuse()` computes `quantize(W + BA)` would train against a
  quantizer that never runs. So this fake-quantizes the **merged** weight, at the base
  layer's own `group_size` / `bits` / `mode`, with a straight-through estimator
  carrying the gradient.

  The load-bearing test derives its expectation by calling the real `fuse()` rather
  than re-deriving the merge arithmetic, so a bug in the implementation cannot be
  mirrored into the test.

  Implementation notes:
  
  - Applied **after** the adapters exist, matching the CUDA ordering (`llama.py:3464`,
  `vision.py:2043`)
  - Uses `__class__` reassignment preserving `__name__`/`__qualname__`, following the
  NEFTune precedent at `trainer.py:3899` — the save path's DoRA detection keys on
  `type(module).__name__` and `collect_mlx_lora_adapter_tensors` relies on it
  - `qat_scheme="auto"` inherits the base quantization; `"int4"`/`"int8"` additionally
  assert the base matches

  ## Scope

  Covers **LoRA over `nn.QuantizedLinear` on text models**. Everything else is refused
  rather than approximated, because each merges differently and a fake-quant that
  doesn't match `fuse()` is worse than none:

  | rejected | reason |
  |---|---|
  | DoRA | `fuse()` rescales by `m / ‖W + BA‖` **before** quantizing |
  | MoE / SwitchLinear | 3-D per-expert weight, different delta layout |
  | VLMs | merge path not yet validated end to end (coverage gate, not a known
  incompatibility) |
  | unquantized base | `fuse()` only requantizes when the base is quantized — nothing
  to simulate |
  | `lora_dropout > 0` | folding the adapter into the weight leaves no separate
  activation path |
  | `int8-int4`, `fp8-int4`, `fp8-fp8`, `cactus` | torchao schemes; quantize
  activations and/or use fp8 grids, not expressible with `mx.quantize` |
  | `full_finetuning=True` | needs a quantize-at-save path that does not exist yet |

  QAT + CPT `modules_to_save` is structurally impossible and asserted as such: CPT
  full modules require an unquantized base, QAT requires a quantized one.

  ## Tests

  31 new tests, registered in `mlx-ci.yml`.

  - `tests/test_mlx_qat.py` (23) — fuse parity with and without bias, STE Jacobian,
  apply/remove lifecycle, class-name preservation, and every rejection above
  - `tests/test_mlx_qat_metal.py` (2) — efficacy: trains with and without QAT from one
  seed, fuses both, asserts the saved model is better. Parametrized over seeds; 400
  steps because the margin is genuinely marginal at 120 (measured minimum across seeds
  0–3: 120 steps −0.000030, 400 +0.001303, 900 +0.002600)
  - `tests/test_mlx_qat_integration_metal.py` (6) — the real product paths:
  `merged_4bit` save/reload round trip, QAT vs non-QAT through that path, adapter-only
  export, `MLXTrainer` (verifying QAT survives `mx.compile`), the VLM gate, and the
  CPT exclusion

  No regressions in the existing MLX suite (`trainer_internals` 223, `switch_lora` 23,
  `save_export_regressions` 49, `adapter_metadata_persistence` 31, others green).
  Three suites have pre-existing failures (`module_exports` 2,
  `save_export_edge_cases` 7, `save_lora_adapters_filter` 12) — confirmed identical
  counts against a pristine checkout of `main`; they are the shim-dependent tests that
  fail standalone on real Metal.

  ## Note for reviewers

  **The training loss reads higher with QAT enabled, and that is the feature
  working.** A non-QAT run's pre-fuse loss describes a model configuration that only
  exists in memory and is destroyed by the save. The comparable number is the loss
  *after* `merged_4bit`. The banner printed on activation says this, but it will still
  look wrong on a loss curve.

  Two things worth deciding before you open it: whether to mention that lifting the
  VLM gate is likely test-only work (all 260 layers on Qwen2-VL-2B are already
  discovered correctly), and whether to flag the follow-up sequence — fp16→quantized
  save, then DoRA/MoE QAT — so reviewers know the scope is intentionally staged rather
  than incomplete.

### CODE REVIEW

**Comment by BardiaKoopah** [comment]
@codex review

**Comment by BardiaKoopah** [comment]
@codex review

File: unsloth_zoo/mlx/qat.py [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+        )
+
+    if lora_dropout is not None and float(lora_dropout) > 0.0:
+        raise NotImplementedError(
+            "Unsloth: qat_scheme requires lora_dropout=0 on MLX. QAT folds the "
+            "adapter into the weight to match fuse(), which leaves no separate "
+            f"LoRA activation path for dropout to act on (got {lora_dropout})."
+        )
+
+    return requested_bits
```
[the remark this answers is not in the document]
Comment (BardiaKoopah) [reply unsloth_zoo/mlx/qat.py:376]: Thanks — fixed in https://github.com/unslothai/unsloth-zoo/commit/0734d9a32804203e92060df98efa98524318f5f3.

Verified the report before fixing: loading with load_in_16bit=True and passing qat_scheme="auto" left 168 LoRA layers attached and 4,399,104 trainable parameters behind when the call raised. Exactly the mutated-on-failure problem the preflight in https://github.com/unslothai/unsloth-zoo/commit/30d344de7abd44270d1f38f4d5a9476fef768981 was meant to end — I moved the scheme / full-finetuning / VLM / dropout checks up and left this one behind in the post-LoRA pass.

Whether anything is quantized is knowable before adapters exist, so the check now lives in validate_mlx_qat_request. The same helper serves both phases, since named_modules() still walks into LoRALinear.linear after LoRA is applied.

Two tests added at the get_peft_model level (not the internal helper, so an unreachable guard fails the suite): one for the dropout refusal, one for the unquantized-base refusal, both asserting the model is unchanged afterwards. Worth noting "unchanged" is not "zero trainable parameters" — a freshly loaded MLX model has everything trainable until get_peft_model freezes it, so the assertion compares against the pre-call count.

For the record on the earlier batch: the full_finetuning preflight and non-affine mx.quantize findings were both real and are fixed. The mode-less-QuantizedLinear one I did not act on — mlx_lm's own LoRALinear.fuse() uses linear.mode unguarded and unsloth_zoo/mlx/loader.py has nine pre-existing unguarded .mode references, so if that floor were reachable the existing loader and mlx-lm itself would already break. That looks like a stale declared minimum rather than something this PR introduces.

**Comment by danielhanchen** [comment]
Confirmed qat_scheme still raises NotImplementedError in unsloth_zoo/mlx/loader.py, and the preflight now validates the same modules linear_to_lora_layers wraps so a rejection cannot leave adapters attached. Will get this reviewed.

**Comment by danielhanchen** [comment]
Checked that main still raises NotImplementedError for qat_scheme in unsloth_zoo/mlx/loader.py, and that the fuse parity test derives its expectation from the real fuse() rather than re-deriving the merge, which is the right call. Will get this reviewed.

**Comment by danielhanchen** [comment]
Checked that qat_scheme still raises NotImplementedError in unsloth_zoo/mlx/loader.py on main, and the fake-quant target here matches what LoRALinear.fuse actually requantizes. Will get this reviewed.
